File: src/net/eduard/api/lib/manager/CommandManager.java
package net.eduard.api.lib.manager;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.eduard.api.lib.modules.Extra;
import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.PluginCommand;
import org.bukkit.command.TabCompleter;
import org.bukkit.command.defaults.BukkitCommand;
import org.bukkit.plugin.Plugin;

import net.eduard.api.lib.Mine;
import net.eduard.api.lib.storage.StorageAttributes;

@StorageAttributes(indentificate = true)
public class CommandManager extends EventsManager implements TabCompleter, CommandExecutor {
    private static boolean debug = true;
    private static Map<String, CommandManager> commandsRegistred = new HashMap<>();

    public static CommandManager getCommand(String name) {
        return commandsRegistred.get(name.toLowerCase());
    }

    @StorageAttributes(reference = true)
    private CommandManager parent;
    protected String name;
    private String permission;

    private transient Command customCommand;

    private String usage;
    private String usagePrefix="§cUtilize ";

    private String description = "Este comando faz algo";
    private List<String> aliases = new ArrayList<>();

    private String permissionMessage = Mine.MSG_NO_PERMISSION;

    private Map<String, CommandManager> commands = new HashMap<>();

    public CommandManager() {
        this("");
    }

    public static boolean isDebug() {
        return debug;
    }

    public static void setDebug(boolean debug) {
        CommandManager.debug = debug;
    }

    public String autoUsage() {
        if (parent != null) {
            return parent.autoUsage() + " " + name;
        } else {

            return "/" + name + "";
        }
    }

    public String autoPermission() {

        if (parent != null) {
            return parent.autoPermission() + "." + name;
        } else {
            return "command." + name;
        }

    }

    public CommandManager(String name, String... aliases) {
        this.name = name;
        if (name.equals("")) {
            this.name = getCommandName();

        }
        if (aliases != null)
            this.aliases = Arrays.asList(aliases);
    }

    public void broadcast(String message) {
        Mine.broadcast(message, permission);

    }

    public List<String> getAliases() {
        return aliases;
    }

    public PluginCommand getCommand() {
        return Bukkit.getPluginCommand(name);
    }

    public String getCommandName() {

        return getClass().getSimpleName().toLowerCase().replace("sub", "").replace("subcommand", "")
                .replace("comando", "").replace("command", "").replace("cmd", "").replace("eduard", "");
    }

    public Map<String, CommandManager> getCommands() {
        return commands;
    }

    public String getDescription() {
        return description;
    }

    public String getName() {
        return name;
    }

    public CommandManager getParent() {
        return parent;
    }

    public String getPermission() {
        return permission;
    }

    public String getPermissionMessage() {
        return permissionMessage;
    }

    public String getUsage() {
        return usage;
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {

        {
            CommandManager cmd = this;
            for (int i = 0; i < args.length; i++) {
                String arg = args[i].toLowerCase();
                CommandManager sub = null;
                for (CommandManager subcmd : cmd.getCommands().values()) {
                    if (subcmd.getName().equalsIgnoreCase(arg)) {
                        sub = subcmd;
                    }
                    for (String alias : subcmd.getAliases()) {
                        if (alias.equalsIgnoreCase(arg)) {
                            sub = subcmd;
                        }
                    }
                }
                if (sub == null) {
                    break;

                }
                cmd = sub;
            }
//			sender.sendMessage("permiscao " + cmd.getPermission());
            if (cmd == this) {
                if (args.length == 0) {
                    sender.sendMessage("/" + name + " help");
                } else {
                    sendUsage(sender);
                }
            } else {

                if (sender.hasPermission(cmd.getPermission())) {
                    cmd.onCommand(sender, command, label, args);
                } else
                    sender.sendMessage(cmd.getPermissionMessage());
            }

        }
        return true;

    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        ArrayList<String> vars = new ArrayList<>();

        {
            CommandManager cmd = this;
            for (int i = 0; i < args.length; i++) {
                String arg = args[i].toLowerCase();
                vars.clear();
                CommandManager sub = null;
                for (CommandManager subcmd : cmd.getCommands().values()) {
                    if (sender.hasPermission(subcmd.getPermission())) {
                        if (Mine.startWith(subcmd.getName(), arg)) {
                            vars.add(subcmd.getName());
                        }
                        if (subcmd.getName().equalsIgnoreCase(arg)) {
                            sub = subcmd;
                        }
                        for (String alias : subcmd.getAliases()) {
                            if (Mine.startWith(alias, arg)) {
                                vars.add(alias);
                            }
                            if (alias.equalsIgnoreCase(arg)) {
                                sub = subcmd;
                            }
                        }

                    }

                }
                if (sub == null) {
                    break;
                }
                cmd = sub;
            }
            if (vars.isEmpty()) {


                return null;
            }
//			if (cmd == this) {
//				if (vars.isEmpty()) {
//					return null;
//				}
//				return vars;
//			} else {

//				if (sender.hasPermission(cmd.getPermission())) {
//					return cmd.onTabComplete(sender, command, label, args);
//				}
            return vars;
//			}

        }

    }

    public static void log(String msg) {

        if (debug)
            Mine.console("§bCommandAPI §f" + msg);
    }

    public boolean register() {

        PluginCommand command = Bukkit.getPluginCommand(name);
        if (command == null) {
            log("O comando §a" + name + " §fnao foi registrado na plugin.yml de nenhum Plugin do Servidor");
            return false;
        }
        setPlugin(command.getPlugin());
        if (command.getUsage() != null) {
            if (!command.getUsage().isEmpty()) {
                usage = command.getUsage().replace("<command>", name).replace('&', '§');
            }
        }
        if (usage == null) {
            usage = autoUsage();
        }
        if (command.getPermission() != null) {
            permission = command.getPermission();
        } else if (permission == null) {
            permission = autoPermission();
        }
        if (command.getPermissionMessage() != null) {
            permissionMessage = command.getPermissionMessage().replace('&', '§');
        }
        if (command.getDescription() != null) {
            description = command.getDescription().replace('&', '§');
        }
        // alias não funciona para comandos apenas na plugin.yml ou subcomandos
        if (command.getAliases() != null) {
            aliases = command.getAliases();
        }

        command.setUsage(usage);
        command.setLabel(name);
//		command.setName(name);
        command.setAliases(aliases);
        command.setDescription(description);
        command.setPermission(permission);
        command.setExecutor(this);

        log("O comando §a" + name + " §ffoi registrado para o Plugin §b" + command.getPlugin().getName()
                + "§f pela plugin.yml");
        commandsRegistred.put(name.toLowerCase(), this);
        updateSubs();
        register(getPluginInstance());
        return true;

    }

    public boolean register(CommandManager sub) {
        commands.put(sub.name, sub);
        sub.setParent(this);
        return true;
    }

    public void registerCommand(Plugin plugin) {
        setPlugin(plugin);
        if (usage == null) {
            usage = autoUsage();
        }
        if (permission == null) {
            permission = autoPermission();
        }

        Command command = new Command(name) {

            @Override
            public boolean execute(CommandSender sender, String label, String[] args) {
                try {
                    if (!plugin.isEnabled()) {
                        return false;
                    }
                    if (sender.hasPermission(permission)) {

                        return onCommand(sender, this, label, args);

                    } else {
                        sender.sendMessage(permissionMessage);
                    }
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
                return false;
            }

            @Override
            public List<String> tabComplete(CommandSender sender, String alias, String[] args) throws IllegalArgumentException {

                List<String> vars = onTabComplete(sender, this, getName(), args);
                List<String> superior = super.tabComplete(sender, alias, args);
                ;
                return vars == null ? superior : vars;


            }


        };
        command.setAliases(aliases);
        command.setDescription(description);
        command.setLabel(name);
        // command.setName(name);
        command.setUsage(usage);
        command.setPermissionMessage(permissionMessage);
        command.setPermission(permission);
        log("O comando §a" + name + " §ffoi registrado para o Plugin §b" + plugin.getName()
                + "§f sem plugin.yml");
        commandsRegistred.put(name.toLowerCase(), this);

        updateSubs();
        register(plugin);
        Mine.createCommand(plugin, command);
    }

    public void unregisterCommand() {
        if (getCustomCommand() != null)
            Mine.removeCommand(name);

    }

    @Override
    public Object restore(Map<String, Object> map) {
        // TODO Auto-generated method stub
        return null;
    }

    @Override
    public void store(Map<String, Object> map, Object object) {
        // TODO Auto-generated method stub

    }

    public void sendDescription(CommandSender sender) {
        sender.sendMessage(getDescription());
    }

    public void sendPermissionMessage(CommandSender sender) {
        sender.sendMessage(permissionMessage);
    }

    public void sendUsage(CommandSender sender) {
        if (usagePrefix==null){
            usagePrefix="";
        }
        sender.sendMessage(getUsagePrefix() +getUsage());
    }

    public void setAliases(List<String> aliases) {
        this.aliases = aliases;
    }

    protected void setCommands(Map<String, CommandManager> commands) {
        this.commands = commands;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public static Map<String, CommandManager> getCommandsRegistred() {
        return commandsRegistred;
    }

    public static void setCommandsRegistred(Map<String, CommandManager> commandsRegistred) {
        CommandManager.commandsRegistred = commandsRegistred;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setParent(CommandManager parent) {
        this.parent = parent;
    }

    public void setPermission(String permission) {
        this.permission = permission;
    }

    public void setPermissionMessage(String permissionMessage) {
        this.permissionMessage = permissionMessage;
    }

    public void setUsage(String usage) {
        this.usage = usage;
    }

    public void updateSubs() {
        for (CommandManager sub : commands.values()) {
            if (sub.permission == null)
                sub.permission = sub.autoPermission();

            if (sub.usage == null) {
                sub.usage = sub.autoUsage();
            }
            log("O subcomando §e" + sub.name + " §ffoi registrado no comando §a" + name);
            if (!sub.commands.isEmpty())
                sub.updateSubs();
        }
    }

    public Command getCustomCommand() {
        return customCommand;
    }

    public void setCustomCommand(Command customCommand) {
        this.customCommand = customCommand;
    }

    public String getUsagePrefix() {
        return usagePrefix;
    }
}

File: src/main/java/net/eduard/api/lib/manager/CommandManager.kt
package net.eduard.api.lib.manager

import net.eduard.api.lib.kotlin.formatColors
import net.eduard.api.lib.modules.Extra
import net.eduard.api.lib.modules.Mine
import org.bukkit.Bukkit
import org.bukkit.command.*
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import org.bukkit.plugin.java.JavaPlugin
import java.util.*

open class CommandManager(var name: String, vararg aliases: String) : EventsManager(), TabCompleter, CommandExecutor {

    @Transient
    var parent: CommandManager? = null
        set(value) {
            field = value
            if (usage.isEmpty()) {
                usage = autoUsage()
            }
            if (permission.isEmpty()) {
                permission = autoPermission()
            }
        }

    @Transient
    var commandRegistred: Command? = null

    @Transient
    var subCommands: MutableMap<String, CommandManager> = mutableMapOf()

    open fun playerCommand(player: Player, args: Array<String>) {

    }

    open fun command(sender: CommandSender, args: Array<String>) {
        if (sender is Player) {
            playerCommand(sender, args)
        } else {
            if (args.isEmpty()) {
                sender.sendMessage("/$name help")
            } else {
                sendUsage(sender)
            }
        }
    }

    var usagePrefix: String = DEFAULT_USAGE_PREFIX
    var description = DEFAULT_DESCRIPTION
    var permission: String = DEFAULT_PERMISSION
    var usage: String = ""
    var aliases: MutableList<String> = ArrayList()
    var permissionMessage = Mine.MSG_NO_PERMISSION


    val command: PluginCommand
        get() = Bukkit.getPluginCommand(name)

    init {
        this.aliases = aliases.toMutableList()
    }

    constructor() : this("") {}

    private fun autoUsage(): String {
        return if (parent != null) {
            parent?.autoUsage() + " " + name
        } else {
            "/$name"
        }
    }


    private fun autoPermission(): String {
        return if (parent != null) {
            parent?.autoPermission() + "." + name
        } else {
            "command.$name"
        }
    }


    fun broadcast(message: String) {
        Bukkit.broadcast(message, permission)
    }


    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {

        if (args.isNotEmpty()) {
            val arg = args[0]
            var sub: CommandManager? = null
            for (subcmd in subCommands.values) {
                if (subcmd.name.equals(arg, ignoreCase = true)) {
                    sub = subcmd
                }
                for (alias in subcmd.aliases) {
                    if (alias.equals(arg, ignoreCase = true)) {
                        sub = subcmd
                    }
                }
            }
            if (sub != null) {
                val args2 = args.copyOfRange(1, args.size)
                return sub.onCommand(sender, command, label, args2)
            } else {
                if (sender.hasPermission(permission)) {
                    command(sender, args)
                } else {
                    sender.sendMessage(permissionMessage)
                }
            }
        } else {
            if (sender.hasPermission(permission)) {
                command(sender, args)
            } else {
                sender.sendMessage(permissionMessage)
            }
        }
        return true;

    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>
    ): List<String>? {
        val vars = ArrayList<String>()
        var cmdSelected = this
        for (index in args.indices) {
            val arg = args[index].toLowerCase()
            vars.clear()
            var sub: CommandManager? = null
            for (subCMD in cmdSelected.subCommands.values) {
                if (!sender.hasPermission(subCMD.permission)) continue
                if (Extra.startWith(subCMD.name, arg)) {
                    vars.add(subCMD.name)
                }
                if (subCMD.name.equals(arg, ignoreCase = true)) {
                    sub = subCMD
                }
                for (alias in subCMD.aliases) {
                    if (Extra.startWith(alias, arg)) {
                        vars.add(alias)
                    }
                    if (alias.equals(arg, ignoreCase = true)) {
                        sub = subCMD
                    }
                }
            }
            if (sub == null) {
                break
            }
            cmdSelected = sub
        }
        if (cmdSelected != this) {
            val subTabComplete = cmdSelected.onTabComplete(sender, command, label, args)
            if (subTabComplete != null) {
                vars.addAll(subTabComplete)
            }
        }

        return if (vars.isEmpty()) {
            null
        } else vars
    }

    private fun fixPermissionAndUsage() {
        if (usage.isEmpty()) {
            usage = autoUsage()
        }
        if (permission.isEmpty()) {
            permission = autoPermission()
        }
    }

    fun register(): Boolean {
        fixPermissionAndUsage()
        val command = Bukkit.getPluginCommand(name)
        if (command == null) {
            log("O comando §a$name §fnao foi registrado na plugin.yml de nenhum Plugin do Servidor")
            return false
        }
        plugin = command.plugin as JavaPlugin
        if (command.usage != null) {
            if (command.usage.isNotEmpty()) {
                usage = command.usage
                    .replace("<command>", name, false)
                    .replace('&', '§', false)
            }
        }
        if (command.permission != null) {
            permission = command.permission
        }
        if (command.permissionMessage != null) {
            permissionMessage = command.permissionMessage.formatColors()
        }
        if (command.description != null) {
            description = command.description.formatColors()
        }
        // alias não funciona para comandos apenas na plugin.yml ou subcomandos
        if (command.aliases != null) {
            aliases = command.aliases
        }
        command.usage = usage
        command.label = name
        command.aliases = aliases
        command.description = description
        command.permission = permission
        command.executor = this
        log(
            "O comando §a" + name + " §ffoi registrado para o Plugin §b" + command.plugin.name
                    + "§f pela plugin.yml"
        )
        commandsRegistred[name.toLowerCase()] = this
        updateSubs()
        registerListener(plugin)
        return true

    }

    fun register(sub: CommandManager): Boolean {
        subCommands[sub.name] = sub
        sub.parent = this
        return true
    }

    fun registerCommand(plugin: Plugin) {
        fixPermissionAndUsage()
        this.plugin = plugin as JavaPlugin
        val command = CustomCommand(this)
        command.aliases = aliases
        command.description = description
        command.label = name
        command.setName(name);
        command.usage = usage
        command.permissionMessage = permissionMessage
        command.permission = permission
        log("O comando §a" + name + " §ffoi registrado para o Plugin §b" + plugin.name + "§f sem plugin.yml")

        commandsRegistred[name.toLowerCase()] = this
        commandRegistred = command
        updateSubs()
        registerListener(plugin)
        Bukkit.getScheduler().runTask(plugin) {
            try {
                val registrado = getServerCommandsMap()?.register(plugin.name, command) ?: false
                log("O Comando $name foi registrado com sucesso: $registrado")

            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }

    }


    fun unregisterCommand() {
        val command =  commandRegistred ?: return
        val cmdName = command.name.toLowerCase()
        val pluginName = plugin.name.toLowerCase()
        commandsRegistred.remove(name.toLowerCase())
        commandsRegistred.remove(name)
        command.unregister(getServerCommandsMap())
        val commandsMap = getServerCommandsHashMap()!!
        for (aliase in aliases) {
            val aliasLowercase = aliase.toLowerCase()
            log("Removendo aliase §a$aliase§f do comando §b$cmdName")
            commandsMap.remove(aliasLowercase)
            commandsMap.remove("$pluginName:$aliasLowercase")
        }
        try {
            commandsMap.remove(cmdName)
            commandsMap.remove("$pluginName:$cmdName")
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
        log("Removendo o comando §a$cmdName§f do Plugin §b$pluginName")
    }

    fun sendPermissionMessage(sender: CommandSender) {
        sender.sendMessage(permissionMessage)
    }

    fun sendUsage(sender: CommandSender) {
        sender.sendMessage(usagePrefix + usage)
    }


    private fun updateSubs() {
        for (sub in subCommands.values) {
            sub.parent = this
            log("O subcomando §e" + sub.name + " §ffoi registrado no comando §a" + name)
            if (sub.subCommands.isNotEmpty())
                sub.updateSubs()
        }
    }

    companion object {
        var debugEnabled = true
        var DEFAULT_USAGE_PREFIX = "§cUtilize: "
        var DEFAULT_DESCRIPTION = "Descrição do Comando"
        var DEFAULT_PERMISSION = ""
        val commandsRegistred: MutableMap<String, CommandManager> = HashMap()

        /**
         * @return O Hashmap onde estão registrados todos comandos e aliases
         */
        private fun getServerCommandsMap(): CommandMap? {
            try {

                return Extra.getFieldValue(Bukkit.getServer().pluginManager, "commandMap") as CommandMap
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
            return null
        }

        /**
         * @return O Hashmap onde estão registrados todos comandos e aliases
         */
        private fun getServerCommandsHashMap(): MutableMap<String, Command>? {
            try {
                val simpleCommandMap = getServerCommandsMap()
                val field = SimpleCommandMap::class.java.getDeclaredField("knownCommands")
                field.isAccessible = true
                return field.get(simpleCommandMap) as MutableMap<String, Command>
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
            return null
        }

        fun unregisterCommands(plugin: Plugin) {
            val commands = commandsRegistred.values.filter { it.plugin == plugin }
            commands.forEach {
                it.unregisterCommand()
            }
        }

        fun getCommand(name: String): CommandManager {
            return commandsRegistred[name.toLowerCase()]!!
        }

        fun log(message: String) {
            if (debugEnabled)
                Mine.console("§bCommandAPI §f$message")
        }


    }


    /**
     * Classe para registrar Comando sem a plugin.yml parecido com PluginCommand
     */
    class CustomCommand(val command: CommandManager) : Command(command.name) {
        override fun execute(sender: CommandSender, label: String, args: Array<String>): Boolean {
            try {
                if (!command.plugin.isEnabled) {
                    sender.sendMessage("§cPlugin ${command.plugin.name} foi desabilitado.")
                    return false
                }
                if (sender.hasPermission(permission)) {
                    return command.onCommand(sender, this, label, args)
                } else {
                    command.sendPermissionMessage(sender)
                }
            } catch (ex: Exception) {
                sender.sendMessage("§cUm Erro ocorreu ao executar este comando.")
                ex.printStackTrace()
            }

            return false
        }

        @Throws(IllegalArgumentException::class)
        override fun tabComplete(sender: CommandSender, alias: String, args: Array<String>): List<String>? {

            val tabCompletePossible = command.onTabComplete(sender, this, name, args)
            return tabCompletePossible ?: super.tabComplete(sender, alias, args)


        }
    }


}
